articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -213,10 +213,48 @@ Internally, Reliable Collections serialize your objects using .NET's DataContrac
+Alternatively, you can perform a multi-phase upgrade. 
+1. Upgrade service to a new version that
+    - has both the original V1, and the new V2 version of the data contracts included in the service code package;
+    - registers custom V2 [state serializers](/azure/service-fabric/service-fabric-reliable-services-reliable-collections-serialization#custom-serialization), if needed;
+    - performs all operations on the original, V1 collection using the V1 data contracts.
+2. Upgrade service to a new version that
+    - [creates a new, V2 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.getoraddasync);
+    - performs each add, update and delete operation on first V1 and then V2 collections in a single transaction;
+    - performs read operations on the V1 collection only.
+3. Copy all data from the V1 collection to the V2 collection.
+    - This can be done in a background process by the service version deployed in step 2.
+    - [Enumerate all keys](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary2-2.createkeyenumerableasync)
+      in the V1 collection with the
+      [IsolationLevel.Snapshot](/dotnet/api/microsoft.servicefabric.data.beta.ireliablestatemanager2.createtransaction)
+      to avoid locking it for the entire duration of the copy process.
+    - For each key, use a separate transaction with the [IsolationLevel.ReadRepeatable](/dotnet/api/microsoft.servicefabric.data.beta.isolationlevel) to 
+        - [TryGetValueAsync](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary-2.trygetvalueasync)
+          from the V1 collection.
+        - If the value has already been removed from the V1 collection since the copy process started,
+          the key should be skipped and not resurected in the V2 collection.
+        - [TryAddAsync](/dotnet/api/microsoft.servicefabric.data.collections.ireliabledictionary.tryaddasync)
+          the value to the V2 collection.
+        - If the value has already been added to the V2 collection since the copy process started,
+          the key should be skipped.
+        - The transaction should be committed only if the `TryAddAsync` returns `true`.
+4. Upgrade service to a new version that
+    - performs read operations on the V2 collection only;
+    - still performs each add, update and delete operation on first V1 and then V2 collections to maintain the option of rolling back to V1.
+5. Upgrade service to a new version that
+    - performs all operations on the V2 collection only;
+    - going back to V1 is no longer possible with a service rollback and would require rolling forward with reversed steps 2-4.
+6. Upgrade service a new version that
+    - [removes the V1 collection](/dotnet/api/microsoft.servicefabric.data.ireliablestatemanager.removeasync).
+7. Wait for log truncation.
+    - By default, this happens every 50MB of writes (adds, updates, and removes) to reliable collections.

```

Comment:
Is there a way to force the logs to get truncated?

Reply:
Other than reducing the default checkpoint threshold, no. I don't think so, at least not at the time of last time this came up in #676

Reply:
I don't think so, not without runtime code changes.